src/catchup/CatchupConfiguration.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include <cstdint>
+
+namespace stellar
+{
+
+// Each catchup can be configured by three parameters - last closed ledger,
+// destination ledger and count of ledgers to apply. Minimum value of last
+// closed ledger is '1' which is ledger number used when initializing database.
+// Value of count can be adjusted in different ways during catchup. If applying
+// count ledgers would mean going before the last closed ledger - it is
+// reduced. Otherwise it can be slightly enlarged (by at most checkpoint
+// frequency value) to ensure, that at least count ledgers are applied.
+//
+// Different values of CatchupConfiguration give the same results as old
+// catchup modes:
+// - CATCHUP_MINIMAL - count set to 0
+// - CATCHUP_COMPLETE - count set to UINT32_MAX
+// - CATCHUP_RECENT - count set to recent value
+//
+// Value of destination ledger can be also set to CatchupConfiguration::CURRENT
+// which means that CatchupWork will get latest checkpoint from history archive
+// and catchup to that instead of destination ledger. This is usefull when
+// doing offline commandline catchups with --catchup-complete, --catchup-at,
+// --catchup-to and --catchup-recent arguments to stellar-core.
+//
+// Because of CatchupConfiguration::CURRENT a ResolvedCatchupConfiguration
+// class is also present. It have the same content as this one, except that
+// value of CatchupConfiguration::CURRENT is replaced with number of last
+// checkpoint.
+class CatchupConfiguration
+{
+  public:
+    static const uint32_t CURRENT = 0;
+
+    /**
+     * Preconditions:
+     * * lastClosedLedger > 0
+     * * toLedger == 0 || toLedger >= lastClosedLedger
```

Comment:
I think you want `toLedger == CatchupConfiguration::CURRENT` not `0`

Reply:
now that I think of it, you could also have avoided the use of magic values it would have been better, maybe by having two constructors, and passing CatchupConfiguration early on (right now it's constructed in startCatchup)

Reply:
I create CatchupConfiguration in startCatchup because of the lastClosedLedger that is available inside LedgerManagerImpl. Previous version passed CatchupConfiguration to startCatchup, but there was a lot of repetition with setting the lastClosedLedger in it - it always used the same value from LedgerManagerImpl.

So I moved that and made setting lastClosedLedger responsibility of LedgerManagerImpl, so there is no possibility of passing there an invalid value.